src/common/models/app-settings/app-settings.ts

```diff
@@ -15,247 +15,167 @@
-    }
-
-    let newDataCubes = value.dataCubes.concat();
-    newDataCubes.splice(index, 1);
+export interface SerializedClientAppSettings {
+  version: number;
+  clientTimeout: number;
+  clusters: ClusterJS[];
+  customization: CustomizationJS;
+  dataCubes: DataCubeJS[];
+}
 
-    value.dataCubes = newDataCubes;
-    return new AppSettings(value);
-  }
+export interface ClientAppSettings {
+  version: number;
+  clientTimeout: number;
+  clusters: Cluster[];
+  customization: Customization;
+  dataCubes: DataCube[];
+}
 
-  public attachExecutors(executorFactory: (dataCube: DataCube) => Executor): AppSettings {
-    let value = this.valueOf();
-    value.dataCubes = value.dataCubes.map(ds => {
-      const executor = executorFactory(ds);
-      if (executor) ds = ds.attachExecutor(executor);
-      return ds;
-    });
-    return new AppSettings(value);
-  }
+interface ClustersConfig {
+  clusters?: ClusterJS[];
+  druidHost?: string;
+  brokerHost?: string;
+}
 
-  public getSuggestedCubes(): DataCube[] {
-    return this.dataCubes;
+function readClusters({ clusters, druidHost, brokerHost }: ClustersConfig): Cluster[] {
+  if (Array.isArray(clusters)) return clusters.map(cluster => Cluster.fromJS(cluster));
+  if (isTruthy(druidHost) || isTruthy(brokerHost)) {
+    return [Cluster.fromJS({
+      name: "druid",
+      url: druidHost || brokerHost
+    })];
   }
+  return [];
+}
 
-  public validate(): boolean {
-    return this.customization.validate();
-  }
+interface DataCubesConfig {
+  dataCubes?: DataCubeJS[];
+  dataSources?: DataCubeJS[];
+}
 
-  changeCustomization(customization: Customization): AppSettings {
-    return this.change("customization", customization);
-  }
+function readDataCubes({ dataCubes, dataSources }: DataCubesConfig, clusters: Cluster[]): DataCube[] {
+  const cubes = dataCubes || dataSources || [];
+  return cubes.map(cube => {
+    const cluster = findCluster(cube, clusters);
+    return DataCube.fromJS(cube, { cluster });
+  });
+}
 
-  changeClusters(clusters: Cluster[]): AppSettings {
-    return this.change("clusters", clusters);
-  }
+export function fromConfig(config: AppSettingsYAML): ServerAppSettings {
+  const clusters = readClusters(config);
+  const clientTimeout = config.clientTimeout === undefined ? DEFAULT_CLIENT_TIMEOUT : config.clientTimeout;
+  const dataCubes = readDataCubes(config, clusters);
+  const version = config.version || 0;
+  const customization = Customization.fromJS(config.customization || {});
+
+  // make part of Customization server side smart constructor
+  customization.validate();
+
+  return {
+    clientTimeout,
+    clusters,
+    dataCubes,
+    version,
+    customization
+  };
+}
 
-  addCluster(cluster: Cluster): AppSettings {
-    return this.changeClusters(NamedArray.overrideByName(this.clusters, cluster));
-  }
+export const BLANK_SETTINGS: ServerAppSettings = fromConfig({});
+
+export function toClientSettings({
```

Comment:
I would extract client settings into separate file/module and inverse dependency. E.g client settings should not about server settings.

Reply:
`ServerSettings` are different thing. First we need names, then we can extract if names work.